### PR TITLE
start fractional_ideal

### PR DESCRIPTION
**Pull request by tthsqe12**
This is just the beginning. The real fun will be inverting these.

### CODE REVIEW

**Comment by thofma** [comment]
(@benlorenz I think we saw this Polymake.jl error before at https://github.com/oscar-system/Oscar.jl/pull/404/checks?check_run_id=2299527666 before? Should we just ignore this?)

File: src/Rings/FractionalIdeal.jl [hunk 1]
```diff
@@ -0,0 +1,41 @@
+export fractional_ideal
+
+import Nemo: numerator, denominator, iszero
+
+mutable struct FractionalIdeal{S <: Ideal, T <: RingElem}
+  num::S
+  den::T
+end
+
+function fractional_ideal(n::Ideal, d::RingElem)
```
Comment (thofma) [inline src/Rings/FractionalIdeal.jl:19]: Should check `base_ring(n) === parent(d)`.

**Comment by benlorenz** [comment]
> (@benlorenz I think we saw this Polymake.jl error before at https://github.com/oscar-system/Oscar.jl/pull/404/checks?check_run_id=2299527666 before? Should we just ignore this?)

Yes, until someone can reproduce it more reliably. At some point I ran a loop precompiling Polymake repeatedly but still wasn't able to trigger this error.

**Comment by thofma** [comment]
OK, will do, thanks.

File: src/Rings/FractionalIdeal.jl (hunk 1 above)
Comment (tthsqe12) [reply src/Rings/FractionalIdeal.jl:19]: ok. Is there any reason for
```julia
julia> Oscar.AbstractAlgebra.numerator === Oscar.Nemo.numerator
false
```

Comment (thofma) [reply src/Rings/FractionalIdeal.jl:19]: No, open an issue in Nemo.

**Comment by thofma** [comment]
Out of curiosity, why is the integral basis a fractional ideal and not a pair of of elements and denominator as the docstring says?

**Comment by tthsqe12** [comment]
I don't know. I guess someone would have to change the doc string or the code. @wdecker Is it OK to return a fractional ideal here instead of the pair?

**Comment by thofma** [comment]
FWIW, from the name of the function I would expect elements and not a fractional ideal. It is not even clear how to get the integral basis from the fractional ideal. Probably `gens` on the numerator? Not clear if any generating set must yield a $K[x]$-basis, which is what I was asking for.

**Comment by tthsqe12** [comment]
well, technically, in order to get elements, we would have to return elements of the function field of the curve, which we don't have. These elements are the gens(numerator)/denominator.

**Comment by thofma** [comment]
If these are the elements compromising the integral basis and this function is called `integral_basis`, why don't we return those elements?

**Comment by tthsqe12** [comment]
Because they would have to be elements of the function field of a curve. Are you arguing that this function belongs in the PlaneCurve section?
BTW I do see that putting a basis into an ideal and the expecting gens(ideal) to give that basis is a bit suspect.

**Comment by thofma** [comment]
You wrote "These elements are the gens(numerator)/denominator.". Look like perfectly fine elements to return. I guess the docstring should be adjust to mention that the images of those elements form the basis?

How does returning a fractional ideal help with the problem you mentioned? Does a fractional ideal transport everything into the function field magically?

**Comment by wdecker** [comment]
I think there is a misunderstanding here due to wrong language
which I used in emailing to Dan. I apologize for this.

To explain what is at stake, let me summarize the first two pages
of this paper: https://arxiv.org/abs/1505.05054

An affine domain over a field K is the quotient A = R/I of a polynomial
ring R over K modulo a prime ideal I or, equivalently, a finitely generated
K-algebra which is an integral domain. Its normalization \overline(A)
is the integral closure of A in its quotient field Quot(A). By a finiteness result
of Emmy Noether, \overline(A) is a finite A-module. As such, it may be
represented either as an affine domain again or by giving finitely many
fractions (elements of Quot(A)) with a universal denominator which generate
\overline(A) over A. Both representations are returned by the normalization
function as realized by Dan.

In the case A =Q[x,y]/⟨f⟩ with assumptions as in the docstring we even
know more: The integral closures of A and Q[x] in their common
quotient field Quot(A) = Q(x) coincide. So we may represent \overline(A)
either by generators over A as done by the normalization function or
by generators over Q[x]. For example, consider the case where
f(x,y) = y^3-x^2 defines the standard cusp. Then 1 and y^2/x generate
\overline(A) over A, whereas 1, y, y^2/x generate \overline(A) over Q[x].
In general, a representation of the latter type is obtained from one of the
previous type by multiplying with suitable powers of y.

In fact, there is an additional twist in the case A =Q[x,y]/⟨f⟩:
\overline(A) is a free Q[x]-module of rank deg_y(f) (which is 3
in the example of the standard cusp). By definition, an integral basis
is a set of free generators. Such a basis is obtained from any given set
of generators for \overline(A) over Q[x] by a Hermite normal form
computation.

Summing this up, it should be clear that we need to return fractions
as described in the docstring and not a fractional ideal.

**Comment by thofma** [comment]
Ready to merge @tthsqe12?

**Comment by tthsqe12** [comment]
Yup

**Pull request merged**